File: src/monitor_positions.py
"""monitor_positions.py - watch OPEN trades against stop / target / time-stop using live LTP.

This is a MONITOR/ALERT tool only - it never places or modifies orders. You still
execute manually in your broker. It polls live LTP and prints an ALERT when a position
hits its stop, target, or time-stop.

Setup:
  1. Copy positions.example.json -> positions.json and fill in your open trades.
     (instrument_key must be the tradable contract key, e.g. an NSE_FO option key.)
  2. Ensure a valid .access_token exists.
  3. Run during market hours:
        python src/monitor_positions.py
     Ctrl+C to stop.
"""

import os
import sys
import json
import time
import datetime

import upstox_client
from upstox_client.rest import ApiException

PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
TOKEN_FILE = os.path.join(PROJECT_ROOT, ".access_token")
POSITIONS_FILE = os.path.join(PROJECT_ROOT, "config", "positions.json")
POLL_SECONDS = 10
API_VERSION = "2.0"
IST = datetime.timezone(datetime.timedelta(hours=5, minutes=30))


def load_token():
    if not os.path.exists(TOKEN_FILE):
        print(f"ERROR: {TOKEN_FILE} not found. Authenticate first.")
        sys.exit(1)
    with open(TOKEN_FILE, "r", encoding="utf-8") as f:
        t = f.read().strip()
    if not t:
        print(f"ERROR: {TOKEN_FILE} is empty.")
        sys.exit(1)
    return t


def load_positions():
    if not os.path.exists(POSITIONS_FILE):
        print(f"ERROR: {POSITIONS_FILE} not found.")
        print("Copy config/positions.example.json -> config/positions.json and fill in your trades.")
        sys.exit(1)
    with open(POSITIONS_FILE, "r", encoding="utf-8") as f:
        positions = json.load(f)
    if not isinstance(positions, list) or not positions:
        print(f"ERROR: {POSITIONS_FILE} must be a non-empty JSON list.")
        sys.exit(1)
    return positions


def get_ltps(client, keys):
    """Return {instrument_key: last_price}. Best-effort across SDK response shapes."""
    api = upstox_client.MarketQuoteV3Api(client)
    resp = api.get_ltp(",".join(keys), API_VERSION)
    out = {}
    data = getattr(resp, "data", {}) or {}
    # data is keyed by an exchange symbol; match back by instrument_token when present.
    for _, v in (data.items() if hasattr(data, "items") else []):
        price = getattr(v, "last_price", None)
        token = getattr(v, "instrument_token", None)
        if token is not None and price is not None:
            out[token] = price
    return out, data


def check(position, ltp):
    """Return an alert string if stop/target hit, else None."""
    side = position.get("side", "LONG").upper()
    stop = position.get("stop")
    target = position.get("target")
    if side == "LONG":
        if stop is not None and ltp <= stop:
            return f"STOP hit (LTP {ltp} <= stop {stop})"
        if target is not None and ltp >= target:
            return f"TARGET hit (LTP {ltp} >= target {target})"
    else:  # SHORT
        if stop is not None and ltp >= stop:
            return f"STOP hit (LTP {ltp} >= stop {stop})"
        if target is not None and ltp <= target:
            return f"TARGET hit (LTP {ltp} <= target {target})"
    return None


def time_stop_passed(position, now_ist):
    ts = position.get("time_stop")
    if not ts:
        return False
    try:
        hh, mm = [int(x) for x in ts.split(":")]
        return (now_ist.hour, now_ist.minute) >= (hh, mm)
    except ValueError:
        return False


def main():
    client = upstox_client.ApiClient(_config(load_token()))
    positions = load_positions()
    keys = [p["instrument_key"] for p in positions if p.get("instrument_key")]
    done = set()

    print(f"Monitoring {len(positions)} position(s), polling every {POLL_SECONDS}s. Ctrl+C to stop.\n")
    try:
        while len(done) < len(positions):
            now = datetime.datetime.now(IST)
            try:
                ltps, raw = get_ltps(client, keys)
            except ApiException as e:
                print(f"[{now:%H:%M:%S}] LTP fetch error HTTP {e.status}; retrying...")
                time.sleep(POLL_SECONDS)
                continue

            for idx, p in enumerate(positions):
                if idx in done:
                    continue
                label = p.get("label", p.get("instrument_key", f"pos{idx}"))
                # match LTP: try by instrument_token map, else skip if unavailable
                ltp = None
                tok = p.get("instrument_key")
                # ltps is keyed by instrument_token; many setups key raw by symbol -
                # fall back to scanning raw values for this contract is unreliable, so
                # we report if we cannot resolve.
                for v in (raw.values() if hasattr(raw, "values") else []):
                    if getattr(v, "instrument_token", None) == tok:
                        ltp = getattr(v, "last_price", None)
                        break
                if ltp is None and tok in ltps:
                    ltp = ltps[tok]

                if ltp is None:
                    print(f"[{now:%H:%M:%S}] {label}: LTP unavailable (check instrument_key).")
                    continue

                if time_stop_passed(p, now):
                    print(f"[{now:%H:%M:%S}] ** TIME-STOP ** {label}: exit now (LTP {ltp}).")
                    done.add(idx)
                    continue

                alert = check(p, ltp)
                if alert:
                    print(f"[{now:%H:%M:%S}] ** ALERT ** {label}: {alert} -> ACT NOW.")
                    done.add(idx)
                else:
                    print(f"[{now:%H:%M:%S}] {label}: LTP {ltp} (stop {p.get('stop')}, target {p.get('target')})")

            if len(done) < len(positions):
                time.sleep(POLL_SECONDS)
        print("\nAll positions hit an exit condition. Monitor done.")
    except KeyboardInterrupt:
        print("\nStopped by user.")


def _config(token):
    cfg = upstox_client.Configuration()
    cfg.access_token = token
    return cfg


if __name__ == "__main__":
    main()

File: src/main/java/com/zaheenmunshi/upstox/MonitorPositions.java
package com.zaheenmunshi.upstox;

import com.fasterxml.jackson.core.type.TypeReference;
import com.upstox.ApiClient;
import com.upstox.ApiException;
import com.upstox.api.GetMarketQuoteLastTradedPriceResponseV3;
import com.upstox.api.MarketQuoteSymbolLtpV3;
import io.swagger.client.api.MarketQuoteV3Api;

import java.nio.file.Files;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

/**
 * Port of {@code monitor_positions.py}: watch OPEN trades against stop / target /
 * time-stop using live LTP. MONITOR/ALERT ONLY — it never places or modifies
 * orders; you still execute manually in your broker.
 */
public final class MonitorPositions {

    private static final int POLL_SECONDS = 10;
    private static final ZoneOffset IST = ZoneOffset.ofHoursMinutes(5, 30);
    private static final DateTimeFormatter HMS = DateTimeFormatter.ofPattern("HH:mm:ss");

    private MonitorPositions() {}

    public static void main(String[] args) throws Exception {
        Config cfg = Config.load();
        ApiClient client = UpstoxClients.authenticated(cfg.loadToken());
        MarketQuoteV3Api api = new MarketQuoteV3Api(client);

        List<Map<String, Object>> positions = loadPositions(cfg);
        List<String> keys = new ArrayList<>();
        for (Map<String, Object> p : positions) {
            Object k = p.get("instrument_key");
            if (k != null) keys.add(String.valueOf(k));
        }
        Set<Integer> done = new HashSet<>();

        System.out.println("Monitoring " + positions.size() + " position(s), polling every "
                + POLL_SECONDS + "s. Ctrl+C to stop.\n");

        try {
            while (done.size() < positions.size()) {
                OffsetDateTime now = OffsetDateTime.now(IST);
                Map<String, MarketQuoteSymbolLtpV3> raw;
                try {
                    GetMarketQuoteLastTradedPriceResponseV3 resp = api.getLtp(String.join(",", keys));
                    raw = resp.getData() != null ? resp.getData() : Map.of();
                } catch (ApiException e) {
                    System.out.println("[" + now.format(HMS) + "] LTP fetch error HTTP " + e.getCode() + "; retrying...");
                    Thread.sleep(POLL_SECONDS * 1000L);
                    continue;
                }

                for (int idx = 0; idx < positions.size(); idx++) {
                    if (done.contains(idx)) continue;
                    Map<String, Object> p = positions.get(idx);
                    String label = str(p.getOrDefault("label",
                            p.getOrDefault("instrument_key", "pos" + idx)));
                    String tok = p.get("instrument_key") == null ? null : str(p.get("instrument_key"));

                    Double ltp = null;
                    for (MarketQuoteSymbolLtpV3 v : raw.values()) {
                        if (tok != null && tok.equals(v.getInstrumentToken())) { ltp = v.getLastPrice(); break; }
                    }
                    if (ltp == null && tok != null && raw.containsKey(tok)) {
                        ltp = raw.get(tok).getLastPrice();
                    }

                    if (ltp == null) {
                        System.out.println("[" + now.format(HMS) + "] " + label + ": LTP unavailable (check instrument_key).");
                        continue;
                    }

                    if (timeStopPassed(p, now)) {
                        System.out.println("[" + now.format(HMS) + "] ** TIME-STOP ** " + label + ": exit now (LTP " + ltp + ").");
                        done.add(idx);
                        continue;
                    }

                    String alert = check(p, ltp);
                    if (alert != null) {
                        System.out.println("[" + now.format(HMS) + "] ** ALERT ** " + label + ": " + alert + " -> ACT NOW.");
                        done.add(idx);
                    } else {
                        System.out.println("[" + now.format(HMS) + "] " + label + ": LTP " + ltp
                                + " (stop " + p.get("stop") + ", target " + p.get("target") + ")");
                    }
                }

                if (done.size() < positions.size()) Thread.sleep(POLL_SECONDS * 1000L);
            }
            System.out.println("\nAll positions hit an exit condition. Monitor done.");
        } catch (InterruptedException e) {
            System.out.println("\nStopped by user.");
            Thread.currentThread().interrupt();
        }
    }

    @SuppressWarnings("unchecked")
    private static List<Map<String, Object>> loadPositions(Config cfg) throws Exception {
        var file = cfg.positionsFile();
        if (!Files.exists(file)) {
            System.out.println("ERROR: " + file + " not found.");
            System.out.println("Copy config/positions.example.json -> config/positions.json and fill in your trades.");
            System.exit(1);
        }
        List<Map<String, Object>> positions =
                Json.MAPPER.readValue(file.toFile(), new TypeReference<List<Map<String, Object>>>() {});
        if (positions == null || positions.isEmpty()) {
            System.out.println("ERROR: " + file + " must be a non-empty JSON list.");
            System.exit(1);
        }
        return positions;
    }

    /** Alert string if stop/target hit, else null. */
    private static String check(Map<String, Object> p, double ltp) {
        String side = str(p.getOrDefault("side", "LONG")).toUpperCase();
        Double stop = dbl(p.get("stop"));
        Double target = dbl(p.get("target"));
        if (side.equals("LONG")) {
            if (stop != null && ltp <= stop) return "STOP hit (LTP " + ltp + " <= stop " + stop + ")";
            if (target != null && ltp >= target) return "TARGET hit (LTP " + ltp + " >= target " + target + ")";
        } else {
            if (stop != null && ltp >= stop) return "STOP hit (LTP " + ltp + " >= stop " + stop + ")";
            if (target != null && ltp <= target) return "TARGET hit (LTP " + ltp + " <= target " + target + ")";
        }
        return null;
    }

    private static boolean timeStopPassed(Map<String, Object> p, OffsetDateTime nowIst) {
        Object ts = p.get("time_stop");
        if (ts == null || str(ts).isBlank()) return false;
        try {
            String[] parts = str(ts).split(":");
            int hh = Integer.parseInt(parts[0]), mm = Integer.parseInt(parts[1]);
            return nowIst.getHour() > hh || (nowIst.getHour() == hh && nowIst.getMinute() >= mm);
        } catch (RuntimeException e) {
            return false;
        }
    }

    private static String str(Object o) { return o == null ? null : String.valueOf(o); }

    private static Double dbl(Object o) {
        if (o == null) return null;
        if (o instanceof Number n) return n.doubleValue();
        try { return Double.parseDouble(String.valueOf(o)); } catch (NumberFormatException e) { return null; }
    }
}
